assets/src/edit-story/panels/size/useSize.js

```diff
@@ -0,0 +1,59 @@
+
+/**
+ * WordPress dependencies
+ */
+import { useCallback, useState, useEffect } from '@wordpress/element';
+
+/**
+ * Internal dependencies
+ */
+
+function useSize( width, height ) {
+	const [ lockRatio, setLockRatio ] = useState( true );
+	const [ state, setState ] = useState( { width, height } );
+
+	useEffect( () => {
+		setState( { width, height } );
+	}, [ width, height, setState ] );
+
+	const handleToggleLockRatio = useCallback( () => {
+		setLockRatio( ( val ) => ! val );
+	}, [ setLockRatio ] );
+
+	const handleWidth = useCallback( ( value ) => {
+		const ratio = width / height;
+		const newWidth = isNaN( value ) || value === '' ? '' : parseFloat( value );
+		setState( {
+			...state,
+			width: newWidth,
+			height: typeof newWidth === 'number' && lockRatio ? newWidth / ratio : height,
+		} );
+	}, [ width, height, state, setState, lockRatio ] );
+
+	const handleHeight = useCallback( ( value ) => {
+		const ratio = width / height;
+		const newHeight = isNaN( value ) || value === '' ? '' : parseFloat( value );
+		setState( {
+			...state,
+			height: newHeight,
+			width: typeof newHeight === 'number' && lockRatio ? newHeight * ratio : width,
+		} );
+	}, [ width, height, state, setState, lockRatio ] );
+
+	const handleSubmit = useCallback( ( onSetProperties ) => ( event ) => {
+		onSetProperties( state );
+		event.preventDefault();
+	}, [ state ]	);
```

Comment:
I wasn't sure about moving `handleSubmit` in here. Do you feel it belongs here?

Though pretty elegant with the outer function taking the setter. I could also have been an argument to the custom hook though, as it's not changing - e.g. `function useSize( width, height, onSetProperties ) {`. Which is to be preferred?

Reply:
I think this actually like splitting logic and outer separately. Basically it's focusing on clean output for main component but separating the logic and outer part makes more sense for me.
And onSetProperties on constructor or on function, honestly I'm not sure which is better...
but width and height could be the required input for this useSize. onSetProperties is only for handleSubmit function. So in my opinion it should follow the handleSubmit

Reply:
Thinking a bit more about it, this `handleSubmit` will be identical to (almost*) the other `handleSubmit`'s in all the other panels. Thus if it should be split into a hook, it should be a common hook for all panels. But let's keep it like this for now.

* I think some of the handlers does a little magic before saving, but not sure.

Reply:
Thinking about it again again, let's split the handleSubmit into its own hook in stead. Which we can then generalize better. I don't feel like it belongs in the same hook as these other states and callbacks.

Reply:
yes, it makes sense for me. in case we make this useSize as common util then the handleSubmit could be different for other sections.